File: app/src/main/java/com/aleksandra/go4mytrip/trips/Trips.java
package com.aleksandra.go4mytrip.trips;


import android.content.Intent;
import android.graphics.Canvas;
import android.os.Bundle;
import android.os.Handler;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.PopupMenu;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.aleksandra.go4mytrip.login.MainActivity;
import com.aleksandra.go4mytrip.lists.PackingModel;
import com.aleksandra.go4mytrip.R;
import com.aleksandra.go4mytrip.login.User;
import com.aleksandra.go4mytrip.costestimate.CostEstimateModel;
import com.aleksandra.go4mytrip.notes.NoteModel;
import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Status;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

import it.xabaras.android.recyclerview.swipedecorator.RecyclerViewSwipeDecorator;

public class Trips extends AppCompatActivity implements PopupMenu.OnMenuItemClickListener, TripsListener {
    TextView nameAndSurname;
    TextView emailData;
    ImageView logoutBtn;
    ImageView image2;
    String email;
    String uriPhoto;
    String userName;
    FloatingActionButton addTripBtn;
    List<TripModel> tripList;
    List<CostEstimateModel> costEstimatesList;
    List<NoteModel> noteList;
    List<PackingModel> packingLists;
    DatabaseReference referenceTrips;
    DatabaseReference referenceUser;
    DatabaseReference referencePackingList;
    DatabaseReference referenceExpenses;
    DatabaseReference referenceNotes;
    DatabaseReference referencePlaces;
    FirebaseUser user;
    String uid;
    RecyclerView mrv;
    String nameOfTrip;
    int imageUriOfTrip;
    String dateStart;
    String dateEnd;
    TripModel deletedTrip = null;
    EditText inputSearch;
    ImageView noTrips;
    TextView noTripsText;
    private int tripClickedPosition = -1;
    public static final int REQUEST_CODE_UPDATE_TRIP = 2;
    public static final int REQUEST_CODE_ADD_TRIP = 1;

    int[] images = {R.drawable.img1, R.drawable.img2, R.drawable.img3, R.drawable.img4, R.drawable.img5, R.drawable.img6, R.drawable.img7,
            R.drawable.img8, R.drawable.img9, R.drawable.img10, R.drawable.img11, R.drawable.img12, R.drawable.img13, R.drawable.img14,
            R.drawable.img15, R.drawable.img16, R.drawable.img17, R.drawable.img18, R.drawable.img19, R.drawable.img20,
            R.drawable.img21, R.drawable.img22, R.drawable.img24, R.drawable.img25};

    public static final int REQUEST_CODE = 1;
    GoogleApiClient mGoogleApiClient;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_trips);

       /* nameAndSurname = findViewById(R.id.personData);
        emailData = findViewById(R.id.personEmail);*/
        image2 = findViewById(R.id.image2);
        addTripBtn = findViewById(R.id.circleButton);
        mrv = findViewById(R.id.recyclerview_id);
        tripList = new ArrayList<>();
        costEstimatesList = new ArrayList<>();
        noteList = new ArrayList<>();
        packingLists = new ArrayList<>();
        logoutBtn = findViewById(R.id.logout_google);
        user = FirebaseAuth.getInstance().getCurrentUser();
        assert user != null;
        uid = user.getUid();
        referenceTrips = FirebaseDatabase.getInstance().getReference("trips");
        referenceUser = FirebaseDatabase.getInstance().getReference("users");
        referencePackingList = FirebaseDatabase.getInstance().getReference("packingList");
        referenceExpenses = FirebaseDatabase.getInstance().getReference("expenses");
        referenceNotes = FirebaseDatabase.getInstance().getReference("notes");
        referencePlaces = FirebaseDatabase.getInstance().getReference("placesToVisit");
        inputSearch = findViewById(R.id.inputSearch);
        noTrips = findViewById(R.id.noTrips);
        noTripsText = findViewById(R.id.noTripsText);

       /* LocalBroadcastManager.getInstance(this).registerReceiver(mMessageReceiver,
                new IntentFilter("custom-id"));*/


       /* nameAndSurname.setText(getIntent().getStringExtra("name"));
        emailData.setText(email);
        uriPhoto = getIntent().getStringExtra("photo");
        Picasso.get().load(uriPhoto).into(image2);*/

        inputSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

                referenceTrips.child(uid).addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                        tripList.clear();
                        for (DataSnapshot tripSnapshot : dataSnapshot.getChildren()) {
                            TripModel trip = tripSnapshot.getValue(TripModel.class);
                            if (trip.getTitle().toLowerCase().contains(inputSearch.getText().toString().toLowerCase())) {
                                tripList.add(trip);
                            }

                        }

                        RecyclerTripAdapter myAdapter = new RecyclerTripAdapter(Trips.this, tripList, Trips.this);
                        mrv.setLayoutManager(new LinearLayoutManager(Trips.this));
                        mrv.setHasFixedSize(true);
                        new ItemTouchHelper(itemTouchHelperCallback).attachToRecyclerView(mrv);
                        mrv.setAdapter(myAdapter);

               /* mrecyclerViewAdapter.setOnItemClickListener(new RecyclerViewAdapter.OnItemClickListener() {
                    @Override
                    public void onItemClick(int position) {
                        Toast.makeText(Trips.this, "position: "+ position, Toast.LENGTH_SHORT).show();
                       // referenceTrips.child(uid).child(position)
                    }
                });*/
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {
                    }
                });


            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        addTripBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivityForResult(
                        new Intent(getApplicationContext(), AddNewTrip.class), REQUEST_CODE_ADD_TRIP);

            }
        });
        referenceTrips.child(uid).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                tripList.clear();
                for (DataSnapshot tripSnapshot : dataSnapshot.getChildren()) {
                    TripModel trip = tripSnapshot.getValue(TripModel.class);
                    tripList.add(trip);
                    if (tripList.size() == 0) {
                        mrv.setVisibility(View.GONE);
                        noTrips.setVisibility(View.VISIBLE);
                        noTripsText.setVisibility(View.VISIBLE);
                    }

                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
            }
        });

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_ADD_TRIP) {
            if (resultCode == RESULT_OK) {

                String timeStart, timeEnd;
                nameOfTrip = data.getStringExtra("nameTrip");
                dateStart = data.getStringExtra("date");
                String namePlace = data.getStringExtra("namePlace");
                String coordinate = data.getStringExtra("coordinates");
                if (data.getStringExtra("timeStart") != null) {
                    timeStart = data.getStringExtra("timeStart");
                } else {
                    timeStart = "00:00:00";
                }
                if (data.getStringExtra("timeEnd") != null) {
                    timeEnd = data.getStringExtra("timeEnd");
                } else {
                    timeEnd = "00:00:00";
                }

                dateEnd = data.getStringExtra("endDate");
                imageUriOfTrip = randomImage();

                String id = referenceTrips.push().getKey();

                if (nameOfTrip != null && imageUriOfTrip != 0 && dateStart != null && dateEnd != null) {
                    TripModel trip = new TripModel(id, nameOfTrip, namePlace, coordinate, imageUriOfTrip, dateStart, dateEnd, timeStart, timeEnd);
                    assert id != null;
                    referenceTrips.child(uid).child(id).setValue(trip);

                    Toast.makeText(this, "Trip added", Toast.LENGTH_SHORT).show();


                    clearTripCard();
                } else {
                    Toast.makeText(Trips.this, "Some data is empty, trip was not added", Toast.LENGTH_SHORT).show();
                    clearTripCard();
                }
            }
        } else if (requestCode == REQUEST_CODE_UPDATE_TRIP) {
            if (resultCode == RESULT_OK) {

                String timeStart, timeEnd;
                String id = data.getStringExtra("id");
                nameOfTrip = data.getStringExtra("nameTrip");
                dateStart = data.getStringExtra("date");
                String namePlace = data.getStringExtra("namePlace");
                String coordinate = data.getStringExtra("coordinates");
                if (data.getStringExtra("timeStart") != null) {
                    timeStart = data.getStringExtra("timeStart");
                } else {
                    timeStart = "00:00:00";
                }
                if (data.getStringExtra("timeEnd") != null) {
                    timeEnd = data.getStringExtra("timeEnd");
                } else {
                    timeEnd = "00:00:00";
                }

                dateEnd = data.getStringExtra("endDate");


                Map<String, Object> editItem = new HashMap<>();
                editItem.put("title", nameOfTrip);
                editItem.put("tripDate", dateStart);
                editItem.put("tripDateEnd", dateEnd);
                editItem.put("coordinate", coordinate);
                editItem.put("namePlace", namePlace);
                editItem.put("timeStart", timeStart);
                editItem.put("timeEnd", timeEnd);

                assert id != null;
                referenceTrips.child(uid).child(id).updateChildren(editItem);

            }

        }

    }

    private int randomImage() {

        Random rand = new Random();
        int choice = images[rand.nextInt(images.length)];

        return choice;
    }


    private void clearTripCard() {
        nameOfTrip = null;
        imageUriOfTrip = 0;
        dateStart = null;
        dateEnd = null;
    }

    @Override
    protected void onStart() {
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build();
        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build();
        mGoogleApiClient.connect();
        super.onStart();

        referenceUser.child(uid).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {

                User user = snapshot.getValue(User.class);

                if (user != null) {
                    email = user.getEmail();
                    userName = user.getName();
                    uriPhoto = user.getImageUser();
                    Picasso.get().load(uriPhoto).into(image2);
                }

               /* nameAndSurname.setText(userName);
                emailData.setText(email);*/

            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });

        referenceTrips.child(uid).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                tripList.clear();
                if (dataSnapshot.exists()) {
                    for (DataSnapshot tripSnapshot : dataSnapshot.getChildren()) {
                        TripModel trip = tripSnapshot.getValue(TripModel.class);
                        tripList.add(trip);
                        if (tripList.size() > 0) {
                            mrv.setVisibility(View.VISIBLE);
                            noTrips.setVisibility(View.GONE);
                            noTripsText.setVisibility(View.GONE);
                        }

                    }

                    RecyclerTripAdapter myAdapter = new RecyclerTripAdapter(Trips.this, tripList, Trips.this);
                    mrv.setLayoutManager(new LinearLayoutManager(Trips.this));
                    mrv.setHasFixedSize(true);
                    new ItemTouchHelper(itemTouchHelperCallback).attachToRecyclerView(mrv);
                    mrv.setAdapter(myAdapter);

                } else {
                    mrv.setVisibility(View.GONE);
                    noTrips.setVisibility(View.VISIBLE);
                    noTripsText.setVisibility(View.VISIBLE);
                }

            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
            }
        });
    }


    public void showPopup(View v) {
        PopupMenu popupMenu = new PopupMenu(this, v);
        popupMenu.setOnMenuItemClickListener(this);
        popupMenu.inflate(R.menu.popup_menu);
        popupMenu.show();
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.item1:

                Auth.GoogleSignInApi.signOut(mGoogleApiClient).setResultCallback(
                        new ResultCallback<Status>() {
                            @Override
                            public void onResult(Status status) {
                                // ...
                                Toast.makeText(getApplicationContext(), "Logged Out", Toast.LENGTH_SHORT).show();
                                Intent i = new Intent(getApplicationContext(), MainActivity.class);
                                FirebaseAuth.getInstance().signOut();
                                startActivity(i);
                                finish();

                            }
                        });

        }
        return true;
    }

    ItemTouchHelper.SimpleCallback itemTouchHelperCallback = new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT | ItemTouchHelper.LEFT) {
        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
            return false;
        }


        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
            int position = viewHolder.getAdapterPosition();
            final TripModel trip = tripList.get(position);
            deletedTrip = trip;
            final String id = trip.getTripId();
            referenceTrips.child(uid).child(id).removeValue();


            Snackbar.make(mrv, "Deleted trip " + deletedTrip.getTitle(), Snackbar.LENGTH_LONG)
                    .setAction("Undo", new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            referenceTrips.child(uid).child(id).setValue(deletedTrip);

                        }
                    }).show();
            Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                public void run() {
                    referenceTrips.child(uid).child(id).addListenerForSingleValueEvent(new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot snapshot) {
                            if (!snapshot.exists()) {
                                referenceExpenses.child(uid).child(id).removeValue();
                                referenceNotes.child(uid).child(id).removeValue();
                                referencePackingList.child(uid).child("trips").child(id).removeValue();
                                referencePlaces.child(uid).child(id).removeValue();
                            }
                        }


                        @Override
                        public void onCancelled(@NonNull DatabaseError error) {

                        }
                    });
                }
            }, 8000);   //8 seconds
        }


        @Override
        public void onChildDraw(@NonNull Canvas c, @NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
            new RecyclerViewSwipeDecorator.Builder(Trips.this, c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
                    .addBackgroundColor(ContextCompat.getColor(Trips.this, R.color.white))
                    .addActionIcon(R.drawable.ic_baseline_delete_24_blue)
                    .create()
                    .decorate();

            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
        }
    };

    @Override
    public void onLongClicked(TripModel tripModel, int position) {
        tripClickedPosition = position;
        String id = tripModel.getTripId();
        String title = tripModel.getTitle();
        String coordinate = tripModel.getCoordinate();
        String tripDate = tripModel.getTripDate();
        String timeStart = tripModel.getTimeStart();
        String timeEnd = tripModel.getTimeEnd();
        int image = tripModel.getImageTrip();
        String tripDateEnd = tripModel.getTripDateEnd();
        String namePlace = tripModel.getNamePlace();

        Intent intent = new Intent(Trips.this, AddNewTrip.class);
        intent.putExtra("isViewOrUpdate", true);
        intent.putExtra("id", id);
        intent.putExtra("title", title);
        intent.putExtra("coordinate", coordinate);
        intent.putExtra("tripDate", tripDate);
        intent.putExtra("timeStart", timeStart);
        intent.putExtra("timeEnd", timeEnd);
        intent.putExtra("image", image);
        intent.putExtra("tripDateEnd", tripDateEnd);
        intent.putExtra("namePlace", namePlace);

        startActivityForResult(intent, REQUEST_CODE_UPDATE_TRIP);
    }
}

File: app/src/main/java/com/aleksandra/go4mytrip/trips/Trips.kt
package com.aleksandra.go4mytrip.trips

import android.content.Intent
import android.graphics.Canvas
import android.os.Bundle
import android.os.Handler
import android.text.Editable
import android.text.TextWatcher
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.aleksandra.go4mytrip.R
import com.aleksandra.go4mytrip.costestimate.CostEstimateModel
import com.aleksandra.go4mytrip.lists.PackingModel
import com.aleksandra.go4mytrip.login.MainActivity
import com.aleksandra.go4mytrip.login.UserModel
import com.aleksandra.go4mytrip.notes.NoteModel
import com.aleksandra.go4mytrip.trips.TripsListener
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import com.squareup.picasso.Picasso
import it.xabaras.android.recyclerview.swipedecorator.RecyclerViewSwipeDecorator
import kotlinx.android.synthetic.main.activity_trips.*
import java.util.*

class Trips : AppCompatActivity(), PopupMenu.OnMenuItemClickListener, TripsListener {
    private var image2: ImageView? = null
    private var email: String? = null
    private var uriPhoto: String? = null
    private var userName: String? = null

    private lateinit var namePlace: String
    private lateinit var coordinate: String
    private var timeStart: String? = null
    private var timeEnd: String? = null

    private lateinit var tripList: MutableList<TripModel>
    private lateinit var costEstimatesList: List<CostEstimateModel>
    private lateinit var noteList: List<NoteModel>
    private lateinit var packingLists: List<PackingModel>
    private lateinit var referenceTrips: DatabaseReference
    private lateinit var referenceUser: DatabaseReference
    private lateinit var referencePackingList: DatabaseReference
    private lateinit var referenceExpenses: DatabaseReference
    private lateinit var referenceNotes: DatabaseReference
    private lateinit var referencePlaces: DatabaseReference
    private lateinit var user: FirebaseUser
    private lateinit var uid: String
    private var nameOfTrip: String? = null
    private var imageUriOfTrip = 0
    private var dateStart: String? = null
    private var dateEnd: String? = null
    private lateinit var deletedTrip: TripModel
    private var tripClickedPosition = -1
    private var images = intArrayOf(R.drawable.img1, R.drawable.img2, R.drawable.img3, R.drawable.img4, R.drawable.img5, R.drawable.img6, R.drawable.img7,
            R.drawable.img8, R.drawable.img9, R.drawable.img10, R.drawable.img11, R.drawable.img12, R.drawable.img13, R.drawable.img14,
            R.drawable.img15, R.drawable.img16, R.drawable.img17, R.drawable.img18, R.drawable.img19, R.drawable.img20,
            R.drawable.img21, R.drawable.img22, R.drawable.img24, R.drawable.img25)

    private lateinit var mGoogleApiClient: GoogleApiClient

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_trips)

        image2 = findViewById(R.id.image2)

        tripList = mutableListOf()
        costEstimatesList = ArrayList()
        noteList = ArrayList()
        packingLists = ArrayList()

        val user = FirebaseAuth.getInstance().currentUser

        user?.let {
            uid = user.uid
        }
        referenceTrips = FirebaseDatabase.getInstance().getReference("trips")
        referenceUser = FirebaseDatabase.getInstance().getReference("users")
        referencePackingList = FirebaseDatabase.getInstance().getReference("packingList")
        referenceExpenses = FirebaseDatabase.getInstance().getReference("expenses")
        referenceNotes = FirebaseDatabase.getInstance().getReference("notes")
        referencePlaces = FirebaseDatabase.getInstance().getReference("placesToVisit")

        inputSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                referenceTrips.child(uid).addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        tripList.clear()
                        dataSnapshot.children.forEach {
                            val trip = it.getValue(TripModel::class.java)
                            trip?.let { it1 ->
                                it1.title?.let { it2 ->
                                    if (it2.toLowerCase(Locale.ROOT).contains(inputSearch.text.toString().toLowerCase(Locale.ROOT))) {
                                        tripList.add(it1)
                                    }
                                }

                            }
                        }
                        val myAdapter = TripAdapter(this@Trips, tripList, this@Trips)
                        recyclerview_id.layoutManager = LinearLayoutManager(this@Trips)
                        recyclerview_id.setHasFixedSize(true)
                        ItemTouchHelper(itemTouchHelperCallback).attachToRecyclerView(recyclerview_id)
                        recyclerview_id.adapter = myAdapter
                    }

                    override fun onCancelled(error: DatabaseError) {}
                })
            }

            override fun afterTextChanged(s: Editable) {}
        })
        circleButton.setOnClickListener {
            startActivityForResult(
                    Intent(applicationContext, AddNewTrip::class.java), REQUEST_CODE_ADD_TRIP)
        }
        referenceTrips.child(uid).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                tripList.clear()
                dataSnapshot.children.forEach {
                    val trip = it.getValue(TripModel::class.java)
                    trip?.let {
                        tripList.add(trip)
                    }
                    if (tripList.size == 0) {
                        recyclerview_id.visibility = View.GONE
                        noTrips.visibility = View.VISIBLE
                        noTripsText.visibility = View.VISIBLE
                    }
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE_ADD_TRIP) {
            if (resultCode == RESULT_OK) {

                data?.let {

                    nameOfTrip = it.getStringExtra("nameTrip")
                    dateStart = it.getStringExtra("date")
                    namePlace = it.getStringExtra("namePlace").toString()
                    coordinate = it.getStringExtra("coordinates").toString()
                    timeStart = if (it.getStringExtra("timeStart") != null) {
                        it.getStringExtra("timeStart")

                    } else {
                        "00:00:00"
                    }

                    timeEnd = if (it.getStringExtra("timeEnd") != null) {
                        it.getStringExtra("timeEnd")
                    } else {
                        "00:00:00"
                    }
                    dateEnd = it.getStringExtra("endDate")
                }

                imageUriOfTrip = randomImage()
                val id = referenceTrips.push().key
                if (imageUriOfTrip != 0 && dateStart != null && dateEnd != null) {
                    val trip = TripModel(id, nameOfTrip, namePlace, coordinate, imageUriOfTrip, dateStart, dateEnd, timeStart, timeEnd)

                    id?.let { referenceTrips.child(uid).child(it).setValue(trip) }
                    Toast.makeText(this, "Trip added", Toast.LENGTH_SHORT).show()
                    clearTripCard()
                } else {
                    Toast.makeText(this@Trips, "Some data is empty, trip was not added", Toast.LENGTH_SHORT).show()
                    clearTripCard()
                }
            }
        } else if (requestCode == REQUEST_CODE_UPDATE_TRIP) {
            if (resultCode == RESULT_OK) {
                data?.let {

                    val tripModelU = it.getParcelableExtra<TripModel>(DATA) as TripModel
                    if (tripModelU.timeStart == null) {
                        tripModelU.timeStart = "00:00:00"
                    }
                    if (tripModelU.timeEnd == null) {
                        tripModelU.timeStart = "00:00:00"
                    }

                    val editItem: MutableMap<String, Any?> = HashMap()
                    editItem["title"] = tripModelU.title
                    editItem["tripDate"] = tripModelU.tripDate
                    editItem["tripDateEnd"] = tripModelU.tripDateEnd
                    editItem["coordinate"] = tripModelU.coordinate
                    editItem["namePlace"] = tripModelU.namePlace
                    editItem["timeStart"] = tripModelU.timeStart
                    editItem["timeEnd"] = tripModelU.timeEnd
                    tripModelU.tripId?.let { it1 -> referenceTrips.child(uid).child(it1).updateChildren(editItem) }

                }
            }
        }
    }

    private fun randomImage(): Int {
        val rand = Random()
        return images[rand.nextInt(images.size)]
    }

    private fun clearTripCard() {
        nameOfTrip = null
        imageUriOfTrip = 0
        dateStart = null
        dateEnd = null
    }

    override fun onStart() {
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build()
        mGoogleApiClient = GoogleApiClient.Builder(this)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build()
        mGoogleApiClient.connect()
        super.onStart()
        referenceUser.child(uid).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val user = snapshot.getValue(UserModel::class.java)
                user?.let {
                    email = it.email
                    userName = it.name
                    uriPhoto = it.imageUser
                    Picasso.get().load(uriPhoto).into(image2)
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
        referenceTrips.child(uid).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                tripList.clear()
                if (dataSnapshot.exists()) {
                    dataSnapshot.children.forEach {
                        val trip = it.getValue(TripModel::class.java)
                        trip?.let { it1 ->
                            tripList.add(it1)
                        }
                        if (tripList.size > 0) {
                            recyclerview_id.visibility = View.VISIBLE
                            noTrips.visibility = View.GONE
                            noTripsText.visibility = View.GONE
                        }
                    }
                    val myAdapter = TripAdapter(this@Trips, tripList, this@Trips)
                    recyclerview_id.layoutManager = LinearLayoutManager(this@Trips)
                    recyclerview_id.setHasFixedSize(true)
                    ItemTouchHelper(itemTouchHelperCallback).attachToRecyclerView(recyclerview_id)
                    recyclerview_id.adapter = myAdapter
                } else {
                    recyclerview_id.visibility = View.GONE
                    noTrips.visibility = View.VISIBLE
                    noTripsText.visibility = View.VISIBLE
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    fun showPopup(v: View) {
        val popupMenu = PopupMenu(this, v)
        popupMenu.setOnMenuItemClickListener(this)
        popupMenu.inflate(R.menu.popup_menu)
        popupMenu.show()
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.item1 -> Auth.GoogleSignInApi.signOut(mGoogleApiClient).setResultCallback { // ...
                Toast.makeText(applicationContext, "Logged Out", Toast.LENGTH_SHORT).show()
                val i = Intent(applicationContext, MainActivity::class.java)
                FirebaseAuth.getInstance().signOut()
                startActivity(i)
                finish()
            }
        }
        return true
    }

    var itemTouchHelperCallback: ItemTouchHelper.SimpleCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT or ItemTouchHelper.LEFT) {
        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            return false
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            val position = viewHolder.adapterPosition
            val trip = tripList[position]
            deletedTrip = trip
            val id = trip.tripId

            id?.let {
                referenceTrips.child(uid).child(it).removeValue()
                Snackbar.make(recyclerview_id, "Deleted trip " + deletedTrip.title, Snackbar.LENGTH_LONG)
                        .setAction("Undo") { referenceTrips.child(uid).child(id).setValue(deletedTrip) }.show()
                val handler = Handler()
                handler.postDelayed({
                    referenceTrips.child(uid).child(it).addListenerForSingleValueEvent(object : ValueEventListener {
                        override fun onDataChange(snapshot: DataSnapshot) {
                            if (!snapshot.exists()) {
                                referenceExpenses.child(uid).child(it).removeValue()
                                referenceNotes.child(uid).child(it).removeValue()
                                referencePackingList.child(uid).child("trips").child(it).removeValue()
                                referencePlaces.child(uid).child(it).removeValue()
                            }
                        }

                        override fun onCancelled(error: DatabaseError) {}
                    })
                }, 8000) //8 seconds
            }


        }

        override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
            RecyclerViewSwipeDecorator.Builder(this@Trips, c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
                    .addBackgroundColor(ContextCompat.getColor(this@Trips, R.color.white))
                    .addActionIcon(R.drawable.ic_baseline_delete_24_blue)
                    .create()
                    .decorate()
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
        }
    }

    override fun onLongClicked(tripModel: TripModel, position: Int) {
        tripClickedPosition = position
        var id: String?
        var title: String?
        var coordinate: String?
        var tripDate: String?
        var timeStart: String?
        var timeEnd: String?
        var image: Int
        var tripDateEnd: String?
        var namePlace: String?

        tripModel.let {
            id = it.tripId
            title = it.title
            coordinate = it.coordinate.toString()
            tripDate = it.tripDate
            timeStart = it.timeStart.toString()
            timeEnd = it.timeEnd.toString()
            image = it.imageTrip
            tripDateEnd = it.tripDateEnd.toString()
            namePlace = it.namePlace.toString()
        }
        val intent = Intent(this@Trips, AddNewTrip::class.java)
        intent.putExtra(AddNewTrip.TRIP, tripModel)
        intent.putExtra("isViewOrUpdate", true)
        startActivityForResult(intent, REQUEST_CODE_UPDATE_TRIP)
    }

    companion object {
        const val REQUEST_CODE_UPDATE_TRIP = 2
        const val REQUEST_CODE_ADD_TRIP = 1
        const val DATA = "DATA"
    }


}